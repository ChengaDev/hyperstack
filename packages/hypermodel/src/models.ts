import glob from 'glob'
import { Sequelize } from 'sequelize-typescript'
import L from 'lodash'
import type { ModelsOpts } from './types'
const load = (patt: string) =>
  glob.sync(patt).map((f) => {
    let mod = null
    try {
      mod = require(f)
    } catch (e: any) {
      console.log(`error dynamically loading '${f}'\n${e.toString()}`) // eslint-disable-line no-console
      throw e
    }

    return mod.default || mod
  })
const isTrue = (val: any) => val === true || val === 'true'

const POOL_MAX_DEFAULT = 5
const POOL_MIN_DEFAULT = 1
const POOL_ACQUIRE_DEFAULT = 30000
const POOL_IDLE_DEFAULT = 10000

const createSequelizeConfig = async ({
  logger,
  config,
  modelsRoot,
}: Pick<ModelsOpts, 'logger' | 'config' | 'modelsRoot'>) => {
  const models = L.reduce(load(modelsRoot), L.merge)
  const max_pool_connection = config.pool?.max || POOL_MAX_DEFAULT
  let min_pool_connection = config.pool?.min || POOL_MIN_DEFAULT

  if (min_pool_connection > max_pool_connection) {
    logger.error(
      `max_pool_connection: ${max_pool_connection} is smaller than min_pool_connection: ${min_pool_connection}. setting min=max`
    )
    min_pool_connection = max_pool_connection
  }

  const uri = config.uri
  return {
    uri,
    models,
    sequelizeConfig: {
      logging: config.logging,
      benchmark: isTrue(config.logging),
      native: config.native,
      ssl: config.ssl,
      models: L.values(models), // or [Player, Team],
      pool: {
        max: max_pool_connection,
        min: min_pool_connection,
        acquire: config.pool?.acquire || POOL_ACQUIRE_DEFAULT,
        idle: config.pool?.idle || POOL_IDLE_DEFAULT,
      },
    },
  }
}

const buildModels = async ({
  logger,
  config,
  modelsRoot,
}: Pick<ModelsOpts, 'logger' | 'config' | 'modelsRoot'>) => {
  const { uri, sequelizeConfig, models } = await createSequelizeConfig({
    logger,
    config,
    modelsRoot,
  })
  const connection = new Sequelize(uri, sequelizeConfig)
  return { models, connection }
}

export { buildModels }
