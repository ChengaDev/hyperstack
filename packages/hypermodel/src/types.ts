import type { Sequelize } from 'sequelize-typescript'
import type { DatabaseOptions, Logger } from '@hyperstackjs/typings'
export { OrderItem } from 'sequelize'
export interface ModelsConvergeConfig {
  dropSchema: boolean
  synchronize: boolean
  migrate: boolean
  truncate: boolean
}
export interface ModelsOpts {
  logger: Logger
  connection: Sequelize
  models: ArrayLike<any>
  config: DatabaseOptions
  migrationsRoot: string
  modelsRoot: string
}
