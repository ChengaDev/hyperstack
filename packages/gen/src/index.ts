import path from 'path'
import { Logger, runner } from 'hygen'
import execa from 'execa'
import { parseAttrs } from './parse-attrs'
const defaultTemplates = path.join(__dirname, 'templates')

export const gen = (argv: string[]) => {
  runner(argv, {
    templates: defaultTemplates,
    cwd: process.cwd(),
    logger: new Logger(console.log.bind(console)), // eslint-disable-line no-console
    debug: !!process.env.DEBUG,
    exec: async (action, _body) => {
      // const opts = body && body.length > 0 ? { input: body } : {}
      const res = await execa(action, {
        shell: true,
      })
      return res
    },
    createPrompter: () => require('enquirer'),
    helpers: {
      parseAttrs,
    },
  }).then(({ success }) => process.exit(success ? 0 : 1))
}
