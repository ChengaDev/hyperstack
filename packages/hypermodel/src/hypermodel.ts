/*
eslint-disable max-classes-per-file
*/

import {
  AllowNull,
  Column,
  DataType,
  Default,
  IsUUID,
  Model,
  PrimaryKey,
} from 'sequelize-typescript'
import type { CountOptions } from 'sequelize/types'

class HyperModel<T> extends Model<T> {
  static async exists(options?: CountOptions) {
    const c = await this.count(options)
    return c !== 0
  }
}
class HyperModelError extends Error {
  readonly isModelError = true

  constructor(message?: string) {
    super(message)
    Object.setPrototypeOf(this, new.target.prototype)
  }
}
class ModelAssertionError extends Error {
  readonly isAssertionError = true

  constructor(message?: string) {
    super(message)
    Object.setPrototypeOf(this, new.target.prototype)
  }
}

export { HyperModel, HyperModelError, ModelAssertionError }

export function UUIDPK(t: any, prop: string) {
  Column(t, prop)
  PrimaryKey(t, prop)
  IsUUID(4)(t, prop)
  AllowNull(false)(t, prop)
  Default(DataType.UUIDV4)(t, prop)
}
