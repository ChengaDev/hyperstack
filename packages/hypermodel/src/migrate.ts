/* eslint-disable @typescript-eslint/no-var-requires */
import { BOOLEAN, DATE, INTEGER, JSON, STRING, TEXT } from 'sequelize'

import L from 'lodash'
import createDebug from 'debug'
import type { ModelsOpts } from './types'
const debug = createDebug('hypermodel:migrate')

const tableBuilder = {
  build: (...args: any) => L.reduce(args, L.merge, {}),
  uniqueConstraint: (name: string, fields: any, opts?: any) => ({
    fields,
    type: 'unique',
    name,
    ...opts,
  }),
  intFK: (name: string, toTable: any, toKey: any, opts?: any) => ({
    [name]: {
      type: INTEGER,
      allowNull: false,
      references: {
        model: toTable,
        key: toKey,
      },
      ...opts,
    },
  }),
  intPK: (name: string, opts?: any) => ({
    [name]: {
      type: INTEGER,
      allowNull: false,
      autoIncrement: true,
      primaryKey: true,
      ...opts,
    },
  }),
  json: (name: string, opts?: any) => ({
    [name]: {
      type: JSON,
      ...opts,
    },
  }),
  text: (name: string, opts?: any) => ({
    [name]: {
      type: TEXT,
      ...opts,
    },
  }),
  string: (name: string, opts?: any) => ({
    [name]: {
      type: STRING,
      ...opts,
    },
  }),
  nonNullString: (name: string, opts?: any) => ({
    [name]: {
      type: STRING,
      allowNull: false,
      ...opts,
    },
  }),
  uniqueString: (name: string, opts?: any) => ({
    [name]: {
      type: STRING,
      allowNull: false,
      unique: true,
      ...opts,
    },
  }),
  date: (name: string, opts?: any) => ({
    [name]: {
      type: DATE,
      ...opts,
    },
  }),
  nonNullDate: (name: string, opts?: any) => ({
    [name]: {
      type: DATE,
      allowNull: false,
      ...opts,
    },
  }),
  bool: (name: string, opts?: any) => ({
    [name]: {
      type: BOOLEAN,
      ...opts,
    },
  }),
  int: (name: string, opts?: any) => ({
    [name]: {
      type: INTEGER,
      ...opts,
    },
  }),
  nonNullInt: (name: string, opts?: any) => ({
    [name]: {
      type: INTEGER,
      allowNull: false,
      ...opts,
    },
  }),
  timestamps: (opts?: any) => ({
    createdAt: {
      type: DATE,
      allowNull: false,
      ...opts,
    },
    updatedAt: {
      type: DATE,
      allowNull: false,
      ...opts,
    },
  }),
}

// NOTE: 'auto' is the only operational value, the rest are there for future implementation
// commands which are not 'auto' are assumed to be run from CLI, 'auto' is part of
// auto migration, and is always in process when the app is starting
// the opposite of auto is 'pending'. might want to change it to 'invoke' or something

const logTransform = ({
  event,
  name,
  durationSeconds,
}: {
  event?: string
  name?: string
  durationSeconds?: number
  path?: string
}) => {
  if (event === 'migrating') {
    return `===>    up   | ${name}`
  } else if (event === 'migrated') {
    return `             | ${name} (in ${durationSeconds}s)`
  } else if (event === 'reverting') {
    return `===>    down | ${name}`
  } else if (event === 'reverted') {
    return `             | ${name} (in ${durationSeconds}s)`
  } else {
    return `${event}: ${name} (tbd)`
  }
}
const migrate =
  (cmd: 'auto' | 'pending' | 'reset' | 'to', _id?: string) =>
  async ({
    migrationsRoot,
    logger,
    connection,
  }: Pick<ModelsOpts, 'migrationsRoot' | 'logger' | 'connection'>) => {
    const { Umzug, SequelizeStorage } = require('umzug')
    const migrationsPath = migrationsRoot
    debug('migration root is:', migrationsPath)
    const storage = new SequelizeStorage({
      sequelize: connection,
    })

    const umzug = new Umzug({
      storage,
      logger: { ...logger, info: (msg: any) => logger.info(logTransform(msg)) },
      migrations: {
        glob: migrationsPath,
        resolve: ({ name, path, context }: any) => {
          // Adjust the migration from the new signature to the v2 signature, making easier to upgrade to v3
          const migration = require(path!)
          const newContext = {
            ...context,
            query: connection.getQueryInterface(),
            t: tableBuilder,
          }
          return {
            name,
            up: async () => await migration.up(newContext),
            down: async () => await migration.down(newContext),
          }
        },
      },
    })

    const { yellow, bold } = logger.colors
    const pref = bold(yellow('models:'))

    // Checks migrations and run them if they are not already applied. To keep
    // track of the executed migrations, a table (and sequelize model) called SequelizeMeta
    // will be automatically created (if it doesn't exist already) and parsed.
    logger.info(`${pref} === migrations: start ===`)
    await umzug.up()
    logger.info(`${pref} === migrations: done  ===`)
    if (cmd !== 'auto') {
      // we are running this from CLI, other
      // wise keep the connection because it runs inprocess as part
      // of auto migration
      await connection.close()
    }
  }

export { migrate, tableBuilder }
