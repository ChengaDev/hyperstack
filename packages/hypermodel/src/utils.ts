// Some functions taken from FoalTS

import { pbkdf2, randomBytes, timingSafeEqual } from 'crypto'
import { promisify } from 'util'

import { strictEqual } from 'assert'

export function generateHexToken(size: number) {
  return randomBytes(size).toString('hex')
}

/**
 * Legacy function to hash passwords. Only kept for backward compatibility.
 * @param password
 */
async function parsePassword(password: string): Promise<string> {
  const salt = (await promisify(randomBytes)(16)).toString('hex')
  const iterations = 100000
  const keylen = 64
  const digest = 'sha256'
  const derivedKey = await promisify(pbkdf2)(
    password,
    salt,
    iterations,
    keylen,
    digest
  )
  return `pbkdf2_${digest}$${iterations}$${salt}$${derivedKey.toString('hex')}`
}

/**
 * Hash a password using the PBKDF2 algorithm.
 *
 * Configured to use PBKDF2 + HMAC + SHA256.
 * The result is a 64 byte binary string (or hex if the legacy option is true).
 *
 * The random salt is 16 bytes long.
 * The number of iterations is 150000.
 * The length key is 32 bytes long.
 *
 * @export
 * @param {string} plainTextPassword - The password to hash.
 * @param {{ legacy?: boolean }} [options={}]
 * @returns {Promise<string>} The derived key with the algorithm name, the number of iterations and the salt.
 */
export async function hashPassword(
  plainTextPassword: string,
  options: { legacy?: boolean } = {}
): Promise<string> {
  // apply basic safeguard
  if (plainTextPassword.trim().length < 6) {
    throw new Error('Password invalid: too short')
  }

  if (options.legacy) {
    return parsePassword(plainTextPassword)
  }
  const saltBuffer = await promisify(randomBytes)(16)
  const iterations = 150000
  const keylen = 32
  const digest = 'sha256'
  const derivedKeyBuffer = await promisify(pbkdf2)(
    plainTextPassword,
    saltBuffer,
    iterations,
    keylen,
    digest
  )

  const salt = saltBuffer.toString('base64')
  const derivedKey = derivedKeyBuffer.toString('base64')
  return `pbkdf2_${digest}$${iterations}$${salt}$${derivedKey}`
}

/**
 * Compare a plain text password and a hash to see if they match.
 *
 * @export
 * @param {string} plainTextPassword - The password in clear text.
 * @param {string} passwordHash - The password hash generated by the `hashPassword` function.
 * @param {{ legacy?: boolean }} [options={}]
 * @returns {Promise<boolean>} True if the hash and the password match. False otherwise.
 */
export async function verifyPassword(
  plainTextPassword: string,
  passwordHash: string,
  options: { legacy?: boolean } = {}
): Promise<boolean> {
  const legacy = options.legacy || false
  const [algorithm, iterations, salt, derivedKey] = passwordHash.split('$')

  strictEqual(algorithm, 'pbkdf2_sha256', 'Invalid algorithm.')

  strictEqual(typeof iterations, 'string', 'Invalid password format.')
  strictEqual(typeof salt, 'string', 'Invalid password format.')
  strictEqual(typeof derivedKey, 'string', 'Invalid password format.')
  strictEqual(
    isNaN(parseInt(iterations, 10)),
    false,
    'Invalid password format.'
  )

  const saltBuffer = Buffer.from(salt, legacy ? 'hex' : 'base64')
  const derivedKeyBuffer = Buffer.from(derivedKey, legacy ? 'hex' : 'base64')
  const digest = 'sha256' // TODO: depends on the algorthim var
  const password = await promisify(pbkdf2)(
    plainTextPassword,
    legacy ? saltBuffer.toString('hex') : saltBuffer,
    parseInt(iterations, 10),
    derivedKeyBuffer.length,
    digest
  )
  return timingSafeEqual(password, derivedKeyBuffer)
}
