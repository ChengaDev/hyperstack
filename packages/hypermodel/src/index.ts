import * as st from 'sequelize-typescript'
import { UUIDPK } from './hypermodel'

export const Schema = { ...st, UUIDPK } as typeof st & { UUIDPK: typeof UUIDPK }

export { HyperModel, HyperModelError, ModelAssertionError } from './hypermodel'
export * from './utils'
export { buildModels } from './models'
export * from './migrate'
export * from './converge'
export * from './types'
