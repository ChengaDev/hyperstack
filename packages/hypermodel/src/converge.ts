import createDebug from 'debug'
import { migrate } from './migrate'
import type { ModelsOpts } from './types'
const debug = createDebug('hypermodel:converge')

export const dropSchema = async ({
  logger,
  connection,
}: Pick<ModelsOpts, 'logger' | 'connection'>) => {
  const { yellow, bold } = logger.colors
  logger.info(`${bold(yellow('models:'))} dropping schema`)
  await connection.query(`DROP TABLE IF EXISTS "SequelizeMeta";`)
  await connection.drop()
}
export const syncSchema = async ({
  logger,
  connection,
}: Pick<ModelsOpts, 'logger' | 'connection'>) => {
  const { yellow, bold } = logger.colors
  logger.info(`${bold(yellow('models:'))} syncing schema`)
  await connection.sync({ force: false })
}
export const truncate = async ({
  logger,
  connection,
  models,
}: Pick<ModelsOpts, 'logger' | 'connection' | 'models'>) => {
  const { yellow, bold } = logger.colors
  logger.info(`${bold(yellow('models:'))} truncating tables`)
  for (const model of Object.values(models)) {
    const m = model as any

    await m.truncate({ restartIdentity: true, cascade: true, force: true })
    // https://github.com/sequelize/sequelize/issues/11152
    if (connection.getDialect() === 'sqlite') {
      await connection.query('PRAGMA foreign_keys = OFF;')
      await connection.query(`DELETE FROM ${m.tableName};`)
      await connection.query(`DELETE FROM sqlite_sequence;`)
    }
  }
}

export const converge = async ({
  migrationsRoot,
  models,
  config,
  connection,
  logger,
}: Omit<ModelsOpts, 'modelsRoot'>) => {
  debug('configuration %o', config)
  if (config.dropSchema) {
    await dropSchema({ logger, connection })
  }
  if (config.synchronize) {
    await syncSchema({ logger, connection })
  }
  if (config.migrate) {
    const mig = migrate('auto')
    await mig({ migrationsRoot, connection, logger })
  }
  if (config.truncate) {
    await truncate({ logger, connection, models })
  }
}
