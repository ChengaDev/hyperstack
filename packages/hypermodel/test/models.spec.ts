import path from 'path'
import { buildModels, migrate } from '../src/index'
import { converge } from '../src/converge'

const config = {
  uri: 'sqlite::memory:',
  ssl: false,
  native: false,
  dropSchema: true,
  synchronize: false,
  truncate: false,
  migrate: false,
  logging: false,
}
const log = (...args: any[]) => {
  console.log(...args) // eslint-disable-line no-console
}
const logger = {
  error: log,
  info: log,
  warn: log,
  trace: log,
  colors: {
    yellow: () => {},
    bold: () => {},
  },
}

describe('models', () => {
  it('buildModels', async () => {
    const root = path.join(__dirname, 'fixtures', 'sequelize-base')
    const modelsRoot = path.join(root, 'app', 'models/*.ts')
    const { models, connection } = await buildModels({
      logger,
      config,
      modelsRoot,
    } as any)
    expect(models).toMatchSnapshot()
    expect(connection).toBeTruthy()
    await connection.sync({ force: true })

    const { PopBand } = models
    try {
      await PopBand.create({
        frontman: 'slash',
        email: 'not an email',
      })
      fail('should have thrown')
    } catch (e: any) {
      expect(e.toString()).toMatch(/on email failed/)
    }
    const p = await PopBand.create({
      frontman: 'slash',
      email: 'slash@gnr.com',
    })
    const d = p.get()
    expect(d.createdAt).toBeTruthy()
    expect(d.updatedAt).toBeTruthy()

    d.createdAt = new Date(1981)
    d.updatedAt = new Date(1981)
    expect(d).toMatchSnapshot()
    expect(p.toJSON()).toMatchSnapshot()

    const b = PopBand.build({
      frontman: 'bruce',
      email: 'dickens@ironmaiden.com',
    })
    expect(b.toJSON()).toMatchSnapshot()
  })
  it('migrates', async () => {
    const root = path.join(__dirname, 'fixtures', 'sequelize-base')
    const modelsRoot = path.join(root, 'app', 'models/*.ts')
    const migrationsRoot = path.join(root, 'config', 'db', 'migrate', '*.js')
    const { connection } = await buildModels({
      logger,
      config,
      modelsRoot,
    } as any)
    await migrate('auto')({ migrationsRoot, logger, connection } as any)
    await connection.query(
      `
      INSERT INTO Users (pid, username, password, firstName, lastName, createdAt, updatedAt)
       VALUES ('pid', 'jo', 'pas1', 'jo', 'kash', date(), date());
      `
    )
    expect(
      JSON.stringify(
        await connection.query(
          `
      SELECT * from Users
      `
        )
      )
    ).toMatch(/jo.*kash/)

    await connection.close()
  })

  it('can converge', async () => {
    const config = {
      uri: 'sqlite::memory:',
      ssl: false,
      native: false,
      dropSchema: true, // notice
      synchronize: false, // notice
      truncate: false,
      migrate: true, // notice
      logging: false,
    }
    const root = path.join(__dirname, 'fixtures', 'sequelize-base')
    const modelsRoot = path.join(root, 'app', 'models/*.ts')
    const migrationsRoot = path.join(root, 'config', 'db', 'migrate', '*.js')
    const { connection, models } = await buildModels({
      logger,
      config,
      modelsRoot,
    } as any)
    expect(
      JSON.stringify(
        await connection.query(
          `
      SELECT 
      name
  FROM 
      sqlite_schema
  WHERE 
      type ='table' 
      `
        )
      )
    ).toMatch(/\[\[\],{}]/)
    await converge({
      migrationsRoot,
      models,
      config,
      connection,
      logger,
    } as any)

    expect(connection).toBeTruthy()

    expect(
      JSON.stringify(
        await connection.query(
          `
      SELECT 
      name
  FROM 
      sqlite_schema
  WHERE 
      type ='table' 
      `
        )
      )
    ).toMatch(/Users/)
    await connection.close()
  })
})
