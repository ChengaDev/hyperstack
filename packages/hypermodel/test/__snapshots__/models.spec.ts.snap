// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models buildModels 1`] = `
{
  "PopBand": [Function],
}
`;

exports[`models buildModels 2`] = `
{
  "createdAt": 1970-01-01T00:00:01.981Z,
  "email": "slash@gnr.com",
  "frontman": "slash",
  "id": 1,
  "updatedAt": 1970-01-01T00:00:01.981Z,
}
`;

exports[`models buildModels 3`] = `
{
  "frontman": "slash",
}
`;

exports[`models buildModels 4`] = `
{
  "frontman": "bruce",
}
`;
